// eslint-plugin-css-modules/build/core/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getStyleImportNodeData = exports.getStyleClasses = exports.getPropertyName = exports.getFilePath = exports.getExportPropsMap = exports.getClassesMap = exports.getAST = exports.fileExists = void 0;
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _fp = _interopRequireDefault(require("lodash/fp"));
var _lodash = _interopRequireDefault(require("lodash"));
var _gonzalesPe = _interopRequireDefault(require("gonzales-pe"));
var _traversalUtils = require("./traversalUtils");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
var styleExtensionRegex = /\.(s?css|less)$/;
function dashesCamelCase(str) {
  return str.replace(/-+(\w)/g, function (match, firstLetter) {
    return firstLetter.toUpperCase();
  });
}
var getFilePath = exports.getFilePath = function getFilePath(context, styleFilePath) {
  var settings = context.settings['css-modules'];
  var dirName = _path["default"].dirname(context.getFilename());
  var basePath = settings && settings.basePath ? settings.basePath : '';
  return styleFilePath.startsWith('.') ? _path["default"].resolve(dirName, styleFilePath) : _path["default"].resolve(basePath, styleFilePath);
};
var getPropertyName = exports.getPropertyName = function getPropertyName(node) {
  var propertyName = node.computed
  /*
     square braces eg s['header']
     we won't use node.property.name because it is for cases like
     s[abc] where abc is a variable
   */ ? node.property.value
  /* dot notation, eg s.header */ : node.property.name;

  /*
     skip property names starting with _
     eg. special functions provided
     by css modules like _getCss()
      Tried to just skip function calls, but the parser
     thinks of normal property access like s._getCss and
     function calls like s._getCss() as same.
   */
  if (!propertyName || _lodash["default"].startsWith(propertyName, '_')) {
    return null;
  }
  return propertyName;
};
var getClassesMap = exports.getClassesMap = function getClassesMap(classes, camelCase) {
  var classesMap = {};

  // Unroll the loop because of performance!
  // Remember that this function will run on every lint (e.g.: on file save)
  switch (camelCase) {
    case true:
      _lodash["default"].forIn(classes, function (value, className) {
        classesMap[className] = className;
        classesMap[_lodash["default"].camelCase(className)] = className;
      });
      break;
    case 'dashes':
      _lodash["default"].forIn(classes, function (value, className) {
        classesMap[className] = className;
        classesMap[dashesCamelCase(className)] = className;
      });
      break;
    case 'only':
      _lodash["default"].forIn(classes, function (value, className) {
        classesMap[_lodash["default"].camelCase(className)] = className;
      });
      break;
    case 'dashes-only':
      _lodash["default"].forIn(classes, function (value, className) {
        classesMap[dashesCamelCase(className)] = className;
      });
      break;
    default:
      _lodash["default"].forIn(classes, function (value, className) {
        classesMap[className] = className;
      });
  }
  return classesMap;
};
var getStyleImportNodeData = exports.getStyleImportNodeData = function getStyleImportNodeData(node) {
  // path from which it was imported
  var styleFilePath = _fp["default"].get('source.value')(node);
  if (styleFilePath && styleExtensionRegex.test(styleFilePath)) {
    var importNode = _fp["default"].compose(_fp["default"].find({
      type: 'ImportDefaultSpecifier'
    }), _fp["default"].get('specifiers'))(node);

    // the default imported name
    var importName = _fp["default"].get('local.name')(importNode);
    if (importName) {
      // it had a default import
      return {
        importName: importName,
        styleFilePath: styleFilePath,
        importNode: importNode
      };
    }
  }
};
var fileExists = exports.fileExists = function fileExists(filePath) {
  try {
    // check if file exists
    _fs["default"].statSync(filePath);
    return true;
  } catch (e) {
    return false;
  }
};

/**
 * @returns AST of the parsed file or null if parse failed
 */
var getAST = exports.getAST = function getAST(filePath) {
  var fileContent = _fs["default"].readFileSync(filePath);
  var syntax = _path["default"].extname(filePath).slice(1); // remove leading .

  try {
    return _gonzalesPe["default"].parse(fileContent.toString(), {
      syntax: syntax
    });
  } catch (e) {
    return null;
  }
};
var getStyleClasses = exports.getStyleClasses = function getStyleClasses(ast) {
  /*
     mutates ast by removing :global scopes
   */
  (0, _traversalUtils.eliminateGlobals)(ast);
  var classesMap = (0, _traversalUtils.getRegularClassesMap)(ast);
  var composedClassesMap = (0, _traversalUtils.getComposesClassesMap)(ast);
  var extendClassesMap = (0, _traversalUtils.getExtendClassesMap)(ast);
  var parentSelectorClassesMap = (0, _traversalUtils.getParentSelectorClassesMap)(ast);
  return _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, classesMap), composedClassesMap), extendClassesMap), parentSelectorClassesMap);
};
var getExportPropsMap = exports.getExportPropsMap = function getExportPropsMap(ast) {
  var exportPropsMap = (0, _traversalUtils.getICSSExportPropsMap)(ast);
  return _objectSpread({}, exportPropsMap);
};
