// eslint-plugin-css-modules/build/rules/no-undef-class.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _lodash = _interopRequireDefault(require("lodash"));
var _core = require("../core");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
var _default = exports["default"] = {
  meta: {
    docs: {
      description: 'Checks that you are using the existent css/scss/less classes',
      recommended: true
    },
    schema: [{
      type: 'object',
      properties: {
        camelCase: {
          "enum": [true, 'dashes', 'only', 'dashes-only']
        }
      }
    }]
  },
  create: function create(context) {
    var camelCase = _lodash["default"].get(context, 'options[0].camelCase');

    /*
       maps variable name to property Object
       map = {
         [variableName]: {
           classesMap: { foo: 'foo', fooBar: 'foo-bar', 'foo-bar': 'foo-bar' },
           node: {...}
         }
       }
        example:
       import s from './foo.scss';
       s is variable name
        property Object has two keys
       1. classesMap: an object with propertyName as key and its className as value
       2. node: node that correspond to s (see example above)
     */
    var map = {};
    return {
      ImportDeclaration: function ImportDeclaration(node) {
        var styleImportNodeData = (0, _core.getStyleImportNodeData)(node);
        if (!styleImportNodeData) {
          return;
        }
        var importName = styleImportNodeData.importName,
          styleFilePath = styleImportNodeData.styleFilePath,
          importNode = styleImportNodeData.importNode;
        var styleFileAbsolutePath = (0, _core.getFilePath)(context, styleFilePath);
        var classesMap = {};
        var exportPropsMap = {};
        if ((0, _core.fileExists)(styleFileAbsolutePath)) {
          var ast = (0, _core.getAST)(styleFileAbsolutePath);
          var classes = ast && (0, _core.getStyleClasses)(ast);
          classesMap = classes && (0, _core.getClassesMap)(classes, camelCase);
          exportPropsMap = ast && (0, _core.getExportPropsMap)(ast);
        }

        // this will be used to check if classes are defined
        _lodash["default"].set(map, "".concat(importName, ".classesMap"), classesMap);

        // this will be used to check if :export properties are defined
        _lodash["default"].set(map, "".concat(importName, ".exportPropsMap"), exportPropsMap);

        // save node for reporting unused styles
        _lodash["default"].set(map, "".concat(importName, ".node"), importNode);
      },
      MemberExpression: function MemberExpression(node) {
        /*
           Check if property exists in css/scss file as class
         */

        var objectName = node.object.name;
        var propertyName = (0, _core.getPropertyName)(node, camelCase);
        if (!propertyName) {
          return;
        }
        var classesMap = _lodash["default"].get(map, "".concat(objectName, ".classesMap"));
        var exportPropsMap = _lodash["default"].get(map, "".concat(objectName, ".exportPropsMap"));
        if (classesMap && classesMap[propertyName] == null && exportPropsMap && exportPropsMap[propertyName] == null) {
          context.report(node.property, "Class or exported property '".concat(propertyName, "' not found"));
        }
      }
    };
  }
};
